lib/DXIL/DxilMetadataHelper.cpp

```diff
@@ -594,7 +595,15 @@ void DxilMDHelper::GetDxilResources(const MDOperand &MDO, const MDTuple *&pSRVs,
 
 void DxilMDHelper::EmitDxilResourceBase(const DxilResourceBase &R, Metadata *ppMDVals[]) {
   ppMDVals[kDxilResourceBaseID        ] = Uint32ToConstMD(R.GetID());
-  ppMDVals[kDxilResourceBaseVariable  ] = ValueAsMetadata::get(R.GetGlobalSymbol());
+  Constant *GlobalSymbol = R.GetGlobalSymbol();
+  // For sm66+, global symbol will be mutated into handle type.
+  // Save hlsl type by generate bitcast on global symbol.
+  if (m_pSM->IsSM66Plus()) {
+    Type *HLSLTy = R.GetHLSLType();
+    IRBuilder<> B(m_Ctx);
+    GlobalSymbol = cast<Constant>(B.CreateBitCast(GlobalSymbol, HLSLTy));
```

Comment:
The only thing you get out of using a builder here is the type equality check to skip creating the bitcast.
I think you can just write this instead:
```
if (HLSLTy && HLSLTy != GlobalSymbol->getType())
  GlobalSymbol = cast<Constant>(ConstantExpr::getCast(Instruction::BitCast, GlobalSymbol, HLSLTy));

```